src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -2,49 +2,152 @@
 
 import java.util.List;
 
+import java.util.Objects;
+
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+
+    public MyLinkedList(){
+    }
+
     @Override
     public void add(T value) {
-
+        Node<T> newNode = new Node<>(last, value, null);
+        if (size == 0) {
+            first = newNode;
+        } else {
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> prev = getNode(index).prev;
+        Node<T> next = getNode(index);
+        Node<T> newNode = new Node<>(prev, value, next);
+        if (prev == null) {
+            first = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        if (next == null) {
+            last = newNode;
+        } else {
+            next.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkElementIndex(index);
+        getNode(index).item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return remove(get(index));
     }
 
     @Override
     public T remove(T t) {
+        for (Node<T> i = first; i != null; i = i.next) {
+            if (Objects.equals(i.item, t)) {
+                return unlink(i);
+            }
+        }
         return null;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+    }
+
+    private void checkPositionIndex(int index) {
```

Comment:
redundant method

Reply:
Not at all, it has different top limit, if I check index only less than size, than I cannot add new element in empty list with Index 0 out of bounds Exception (size = 0, and I cheking for index < size), if I check index that may be equal to size, than I recive NullPointerException when try to get element with index = size;